otherlibs/stdune/src/appendable_list.ml

```diff
@@ -1,13 +1,65 @@
-type 'a t = 'a list -> 'a list
-
-let empty k = k
-let singleton x k = x :: k
-let to_list l = l []
-let ( @ ) a b k = a (b k)
-let cons x xs = singleton x @ xs
-
-let rec concat l k =
-  match l with
-  | [] -> k
-  | t :: l -> t (concat l k)
+type 'a t =
+  | Empty
+  | Singleton of 'a
+  | Cons of 'a * 'a t
+  | Append of 'a t * 'a t
+  | List of 'a list
+  | Concat of 'a t list
+
+let empty = Empty
+let singleton x = Singleton x
+
+let ( @ ) a b =
+  match a, b with
+  | Empty, _ -> b
+  | _, Empty -> a
+  | Singleton a, _ -> Cons (a, b)
+  | _, _ -> Append (a, b)
+;;
+
+let cons x xs =
+  match xs with
+  | Empty -> Singleton x
+  | _ -> Cons (x, xs)
+;;
+
+let to_list_rev =
+  let rec loop acc stack =
+    match stack with
+    | [] -> acc
+    | t :: stack ->
+      (match t with
+       | Empty -> loop acc stack
+       | Singleton x -> loop (x :: acc) stack
+       | Cons (x, xs) -> loop (x :: acc) (xs :: stack)
+       | Append (xs, ys) -> loop acc (xs :: ys :: stack)
+       | List xs -> loop (List.rev_append xs acc) stack
+       | Concat [] -> loop acc stack
+       | Concat (x :: xs) -> loop acc (x :: Concat xs :: stack))
+  in
+  fun t -> loop [] [ t ]
+;;
+
+let to_list xs = List.rev (to_list_rev xs)
+
+let concat = function
+  | [] -> Empty
+  | [ x ] -> x
+  | xs -> Concat xs
```

Comment:
The `[ Empty; Empty ]` case doesn't get converted to `Empty`, which makes `is_empty` incorrect.

Reply:
Fixed now.